cloud_functions/gcs_event_based_ingest/README.md

```diff
@@ -0,0 +1,116 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [BigQuery Serverless Ingest](#bigquery-serverless-ingest)
+  - [Tracking Table](#tracking-table)
+  - [Environment Variables](#environment-variables)
+  - [GCS Object Naming Convention](#gcs-object-naming-convention)
+  - [Triggers](#triggers)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# BigQuery Serverless Ingest
+
+Flexible service for performing  BigQuery file loads to existing tables.
+This service handles splitting load jobs when the data volume exceeds
+the BigQuery 15TB load job limit. The goal of the service is to orchestrate
+BigQuery Load Jobs to many bigquery datasets / tables from a single bucket
+providing transparent configuration that is overridable at any level.
+
+![architecture](img/bq-ingest.png)
+
+## Environment Variables
+- `SUCCESS_FILENAME`: Filename to trigger a load (defaults to `_SUCCESS`).
+
+## GCS Object Naming Convention
+### Data Files
+Data should be ingested to a prefix containing destination dataset and table
+like so:
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/${BQ_TABLE_NAME}/*`
+Note, the table prefix can contain multiple sub-prefixes for handling partitions
+or for configuring historical / incremental loads differently.
+
+### Configuration Files
+The Ingestion has many optional configuration files that should live in
+a special `_config/` prefix at the root of the bucket and/or under the dataset
+and/or table and/or under the partition prefixes.
+
+For example if you have the following files:
+```text
+gs://${INGESTION_BUCKET}/_config/load.json`
+gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`
+gs://${INGESTION_BUCKET}/${BQ_DATASET}/${BQ_TABLE_NAME}/_config/load.json
+```
+The json objects will be merged where key conflicts are resolved by config in
+the closest directory to the data.
+If the files contents were like this:
+`gs://${INGESTION_BUCKET}/_config/load.json`:
+```json
+{
+    "sourceFormat": "CSV",
+    "writeDisposition": "WRITE_APPEND",
+    "schemaUpdateOptions": ["ALLOW_FILED_RELAXATION"]
+}
+```
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`:
+```json
+{
+    "sourceFormat": "AVRO",
+}
+```
+`gs://${INGESTION_BUCKET}/${BQ_DATASET}/_config/load.json`:
+```json
+{
+    "writeDisposition": "WRTITE_TRUNCATE",
+}
+```
+
+The result of merging these would be:
+```json
+{
+    "sourceFormat": "AVRO",
+    "writeDisposition": "WRITE_TRUNCATE",
+    "schemaUpdateOptions": ["ALLOW_FILED_RELAXATION"]
+}
+```
+This configuration system gives us the ability to DRY up common defaults but
+override them at whatever level is appropriate as new cases come up.
+
+#### Transformation SQL
```

Comment:
Perhaps a notice that this will utilize the Query slots as opposed to the free load slot tier to execute would be good.